fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/TooltipComponentCallback.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.client.rendering.v1;
+
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.client.gui.tooltip.TooltipComponent;
+import net.minecraft.item.Item;
+import net.minecraft.client.item.TooltipData;
+
+import net.fabricmc.api.EnvType;
+import net.fabricmc.api.Environment;
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+
+/**
+ * Allows registering a mapping from {@link TooltipData} to {@link TooltipComponent}.
+ * This allows custom tooltips for items: first, override {@link Item#getTooltipData} and return a custom {@code TooltipData}.
+ * Second, register a listener to this event and convert the data to your component implementation if it's an instance of your data class.
+ *
+ * <p>Note that failure to map some data to a component will throw an exception,
+ * so make sure that any data you return in {@link Item#getTooltipData} will be handled by one of the callbacks.
+ */
+@Environment(EnvType.CLIENT)
+public interface TooltipComponentCallback {
+	Event<TooltipComponentCallback> EVENT = EventFactory.createArrayBacked(TooltipComponentCallback.class, listeners -> data -> {
+		for (TooltipComponentCallback listener : listeners) {
+			TooltipComponent component = listener.getComponent(data);
+
+			if (component != null) {
+				return component;
+			}
+		}
+
+		return null;
+	});
+
+	/**
+	 * Return the tooltip component for the passed data, or null if none is available.
```

Comment:
This could also use the inline return added in Java 16, which ["provides content for the first sentence of a method's description, and a 'Returns' section"](https://docs.oracle.com/en/java/javase/16/docs/specs/javadoc/doc-comment-spec.html) (instead of just the desc). But this is fine too.
```suggestion
	 * {@return the tooltip component for the passed data, or null if none is available}
```

Reply:
~~Hope Intellij will start displaying this correctly soon.~~

Reply:
It seems to have been fixed in a 2021.1 patch release: https://youtrack.jetbrains.com/issue/IDEA-258058

Reply:
@Technici4n Worth it to make this addition?

Reply:
Don't know, personally I don't particularly like two lines of javadoc saying exactly the same thing.

Reply:
Hmm, we might need to refer to a common javadoc convention at some point. We currently use this tag extensively in yarn; oracle recommends a return tag even if it means there is no body comment; google style recommends what you do. Anyways we can fix that later.